File: rnamake/unittests/residue_unittest.py
import sys
import unittest
import numpy as np
import sys

from rnamake import all_atom, bead
from rnamake import residue_type, settings, util, exceptions, motif_state

import is_equal


class ResidueUnittest(unittest.TestCase):

    def setUp(self):
        self.rts = residue_type.ResidueTypeSet()

        path = settings.UNITTEST_PATH + "resources/res_strs.dat"
        f = open(path)
        lines = f.readlines()
        f.close()

        residues = []
        for l in lines:
            res = all_atom.Residue.from_str(l, self.rts)
            residues.append(res)
        self.residues = residues

    def test_creation(self):
        """
        make sure creating a Residue object does not throw any exceptions
        """
        try:
            #also check name changing works
            rtype = self.rts.get_type("GUA")
            atoms = [all_atom.Atom("O1P", np.array([1, 2, 3]))]
            res = all_atom.Residue(atoms, rtype, "GUA", 1, "A")
        except:
            self.fail("cannot creat Residue object sucessfully")

    def test_get_atom(self):
        gtype = self.rts.get_type("GUA")
        atoms = [
            all_atom.Atom("P", np.array([0, 1, 2]))
        ]
        res = all_atom.Residue(atoms, gtype, "GUA", 1, "A")

        # can find a real atom
        p_atom = res.get_atom("P")
        self.assertIs(p_atom, res.get_atom(index=0))

        # should throw and error if no error exist
        with self.assertRaises(exceptions.ResidueException):
            res.get_atom("P1")

        # invalid num
        with self.assertRaises(exceptions.ResidueException):
            res.get_atom(index=999)

        # valid name but not initialized
        with self.assertRaises(exceptions.ResidueException):
            res.get_atom("OP1")

        with self.assertRaises(exceptions.ResidueException):
            res.get_atom(index=1)

    def test_get_coords(self):
        gtype = self.rts.get_type("GUA")
        atoms = [
            all_atom.Atom("P", np.array([0, 1, 2]))
        ]
        res = all_atom.Residue(atoms, gtype, "GUA", 1, "A")

        # can find a real atom
        p_atom = res.get_atom("P")
        self.assertIs(p_atom, res.get_atom(index=0))

        # should throw and error if no error exist
        with self.assertRaises(exceptions.ResidueException):
            res.get_coords("P1")

        # invalid num
        with self.assertRaises(exceptions.ResidueException):
            res.get_coords(index=999)

        # valid name but not initialized
        with self.assertRaises(exceptions.ResidueException):
            res.get_coords("OP1")

        with self.assertRaises(exceptions.ResidueException):
            res.get_coords(index=1)

    def test_has_atom(self):
        gtype = self.rts.get_type("GUA")
        atoms = [
            all_atom.Atom("P", np.array([0, 1, 2]))
        ]
        res = all_atom.Residue(atoms, gtype, "GUA", 1, "A")

        self.failUnless(res.has_atom("P") == True)
        self.failUnless(res.has_atom("P1") == False)
        self.failUnless(res.has_atom("OP1") == False )
        self.failUnless(res.has_atom(index=0) == True)
        self.failUnless(res.has_atom(index=999) == False)

        res = self.residues[0]
        # first residue has no P
        self.failUnless(res.has_atom("P") == False)
        self.failUnless(res.has_atom("C1'") == True)

    def test_get_str(self):
        res = self.residues[0]
        s = res.get_str()
        res2 = all_atom.Residue.from_str(s, self.rts)

        self.failUnless(is_equal.are_residues_equal(res, res2, check_uuid=0))

    def test_connected_to(self):
        res1 = self.residues[0]
        res2 = self.residues[1]
        res3 = self.residues[2]

        self.failUnless(all_atom.are_residues_connected(res1, res2) == 1)
        self.failUnless(all_atom.are_residues_connected(res2, res1) == -1)
        self.failUnless(all_atom.are_residues_connected(res1, res3) == 0)

    def test_beads(self):
        res1 = self.residues[0]
        res2 = self.residues[10]

        res1.build_beads()
        res2.build_beads()

        self.failIf(res1.get_num_beads() == 0)
        self.failIf(res2.get_num_beads() == 0)

        for b1 in res1.iter_beads():
            for b2 in res2.iter_beads():
                self.failIf(b1.distance(b2) < settings.CLASH_RADIUS)

    def test_copy(self):
        res = self.residues[0]
        copy_res = all_atom.Residue.copy(res)

        self.failUnless(is_equal.are_residues_equal(res, copy_res))

        copy_res = res.get_copy()
        self.failUnless(is_equal.are_residues_equal(res, copy_res))

    def test_move(self):
        r = self.residues[0]
        c1 = r.get_center()

        r.move(np.array([1, 0, 0]))
        c2 = r.get_center()

        self.failUnlessAlmostEqual(1.0, util.distance(c1, c2))

    def test_state(self):
        r = self.residues[1]
        r.build_beads()
        rs = r.get_state()

        self.failUnless(rs.num_beads() == 3)

        s = rs.to_str()
        rs_copy = motif_state.Residue.from_str(s)

        self.failUnless(rs_copy.num_beads() == 3)

        rs_copy = motif_state.Residue.copy(rs)

        self.failUnless(rs_copy.num_beads() == 3)
        self.failUnless(rs_copy == rs)


class BeadUnittest(unittest.TestCase):

    def setUp(self):
        self.b = bead.Bead(np.array([1,1,1]), bead.BeadType.PHOS)

    def test_creation(self):

        with self.assertRaises(exceptions.ResidueException):
            bead.Bead(np.array([1,1,1]), 4)

    def test_copy(self):
        b_copy = bead.Bead.copy(self.b)
        d = b_copy.distance(self.b)
        self.failUnless(d < 0.0001)

    def test_type_name(self):
        name = self.b.get_type_name()
        if name != "PHOSPHATE":
            self.fail("did not get the right bead name")


def main():
    unittest.main()


if __name__ == '__main__':
    main()

File: unittests/structure/residue_unittest.cpp


//headers for testing
#include "../common.hpp"

//RNAMake Headers
#include "structure/residue_type_set.h"
#include "util/bead.h"
#include "base/file_io.h"
#include "base/settings.h"
#include "structure/residue.h"
#include "structure/is_equal.h"
#include "util/uuid.h"


TEST_CASE( "Test Residues for Structure" ) {
    auto rts  = structure::ResidueTypeSet();
    auto path = base::unittest_resource_dir() + "residue/test_str_to_residue.dat";
    auto lines= base::get_lines_from_file(path);
    auto residues  = structure::ResidueOPs();
    for(auto const & l : lines) {
        if(l.size() < 10) { break; } // end of file
        auto r = std::make_shared<structure::Residue>(l, rts);
        residues.push_back(r);
    }
    
    SUBCASE("test getting atoms by name") {
        auto r = residues[0];
        auto name = String("C1'");
        auto a = r->get_atom(name);

        CHECK(a != nullptr);
        CHECK(a.get_name() == "C1'");
        
        REQUIRE_THROWS_AS(r->get_atom("fake"), structure::ResidueException);
    }
    
    SUBCASE("are residues detecting connections properly") {
        auto r1 = residues[0];
        auto r2 = residues[1];
        auto r3 = residues[2];
        
        SUBCASE("connecting from 5' to 3'") { CHECK(r1->connected_to(*r2) == 1);  }
        SUBCASE("should not be connected")  { CHECK(r1->connected_to(*r3) == 0);  }
        SUBCASE("connecting from 3' to 5'") { CHECK(r3->connected_to(*r2) == -1); }
        
    }
    
    SUBCASE("are residues generating steric beads properly") {
        auto r = residues[1];
        auto beads = r->beads_;
        
        SUBCASE("produced the right number of beads") { CHECK(beads.size() == 3); }
        
        CHECK(beads[0].get_type() == util::BeadType::PHOS);
        CHECK(beads[1].get_type() == util::BeadType::SUGAR);
        CHECK(beads[2].get_type() == util::BeadType::BASE);

    }
    
    SUBCASE("are residues copying correctly") {
        auto r = residues[0];
        auto r2 = std::make_shared<structure::Residue>(*r);
        
        CHECK(are_residues_equal(r, r2));
    }
    
    SUBCASE("are residues stringifing correctly") {
        auto r = residues[0];
        auto s = r->get_str();
        auto r2 = std::make_shared<structure::Residue>(s, rts);
        
        SUBCASE("residues should be the same but not have the same id") {
            CHECK(structure::are_residues_equal(r, r2, 0));
        }
    }
    
}


